discojs/src/aggregator/byzantine.ts

```diff
@@ -0,0 +1,185 @@
+        if (threshold > 1 && (!Number.isInteger(threshold)))
+            throw new Error("absolute thresholds must be integral");
+
+        super(roundCutoff, 1);
+        this.#threshold = threshold;
+        this.clippingRadius = clippingRadius
+        this.maxIterations = maxIterations
+
+        if (threshold < 1) {
+            // Throw exception if threshold and thresholdType are conflicting
+            if (thresholdType === 'absolute') {
+                throw new Error(`thresholdType has been set to 'absolute' but choosing threshold=${threshold} implies that thresholdType should be 'relative'.`)
+            }
+            this.#thresholdType = 'relative'
+        }
+        else if (threshold > 1) {
+            // Throw exception if threshold and thresholdType are conflicting
+            if (thresholdType === 'relative') {
+                throw new Error(`thresholdType has been set to 'relative' but choosing threshold=${threshold} implies that thresholdType should be 'absolute'.`)
+            }
+            this.#thresholdType = 'absolute'
+        }
+        // remaining case: threshold == 1
+        else {
+            // Print a warning regarding the default behavior when thresholdType is not specified
+            if (thresholdType === undefined) {
+                // TODO enforce validity by splitting the different threshold types into separate classes instead of warning
+                debug(
+                    "[WARN] Setting the aggregator's threshold to 100% of the nodes' contributions by default. " +
+                    "To instead wait for a single contribution, set thresholdType = 'absolute'"
+                )
+                this.#thresholdType = 'relative'
+            } else {
+                this.#thresholdType = thresholdType
+            }
+        }
+    }
+
+    /** Checks whether the contributions buffer is full. */
+    override isFull(): boolean {
+        // Make sure that we are over the minimum number of participants
+        // if specified
+        if (this.#minNbOfParticipants !== undefined &&
+            this.nodes.size < this.#minNbOfParticipants) return false
+
+        const thresholdValue =
+            this.#thresholdType == 'relative'
+                ? this.#threshold * this.nodes.size
+                : this.#threshold;
+
+        return (this.contributions.get(0)?.size ?? 0) >= thresholdValue;
+    }
+
+    set minNbOfParticipants(minNbOfParticipants: number) {
+        this.#minNbOfParticipants = minNbOfParticipants
+    }
+
+    override _add(nodeId: client.NodeID, contribution: WeightsContainer): void {
+
+        this.log(
+            this.contributions.hasIn([0, nodeId]) ? AggregationStep.UPDATE : AggregationStep.ADD,
+            nodeId,
+        );
+
+        const beta: number = 0.9; // Momentum parameter
+
+        // Apply worker-side momentum
+
+        const prevMomentum: WeightsContainer = this.contributions.getIn([0, nodeId]) as WeightsContainer;
```

Comment:
if it isn't set (such when a node fails to reply in time), it will actually returned `undefined`, it's safer to show it in type

```suggestion
        const prevMomentum = this.contributions.getIn([0, nodeId]) as WeightsContainer | undefined;
```